Sources/AsyncCollections/FlatMap.swift

```diff
@@ -0,0 +1,124 @@
+    /// - Parameter transform: An async  mapping closure. transform accepts an
+    ///     element of this sequence as its parameter and returns a transformed value of
+    ///     the same or of a different type.
+    /// - Returns: An array containing the transformed elements of this sequence.
+    public func asyncFlatMap<T: Sequence>(_ transform: @Sendable (Element) async throws -> T) async rethrows -> [T.Element] {
+        var result: [T.Element] = []
+        for element in self {
+            result.append(contentsOf: try await transform(element))
+        }
+        return result
+    }
+
+    /// Returns an array containing the concatenated results of calling the
+    /// given transformation with each element of this sequence.
+    ///
+    /// Use this method to receive a single-level collection when your
+    /// transformation produces a sequence or collection for each element.
+    ///
+    /// This differs from `asyncFlatMap` in that it uses a `TaskGroup` to run the transform
+    /// closure for all the elements of the Sequence. This allows all the transform closures
+    /// to run concurrently instead of serially. Returns only when the closure has been run
+    /// on all the elements of the Sequence.
+    /// - Parameters:
+    ///   - priority: Task priority for tasks in TaskGroup
+    ///   - transform: An async  mapping closure. transform accepts an
+    ///     element of this sequence as its parameter and returns a transformed value of
+    ///     the same or of a different type.
+    /// - Returns: An array containing the transformed elements of this sequence.
+    public func concurrentFlatMap<T: Sequence>(priority: TaskPriority? = nil, _ transform: @Sendable @escaping (Element) async throws -> T) async rethrows -> [T.Element] where T: Sendable {
+        let result: ContiguousArray<(Int, T)> = try await withThrowingTaskGroup(of: (Int, T).self) { group in
+            for (index, element) in self.enumerated() {
+                group.addTask(priority: priority) {
+                    let result = try await transform(element)
+                    return (index, result)
+                }
+            }
+            // Code for collating results copied from Sequence.map in Swift codebase
+            let initialCapacity = underestimatedCount
+            var result = ContiguousArray<(Int, T)>()
+            result.reserveCapacity(initialCapacity)
+
+            // Add elements up to the initial capacity without checking for regrowth.
+            for _ in 0..<initialCapacity {
+                try await result.append(group.next()!)
+            }
+            // Add remaining elements, if any.
+            while let enumerated = try await group.next() {
+                result.append(enumerated)
+            }
+            return result
+        }
+
+        // construct final array and fill in elements
+        return [T](unsafeUninitializedCapacity: result.count) { buffer, count in
```

Comment:
Can you not calculate the size of the array by adding up the count of all the elements of result and then each array into that buffer. Instead of creating an array of arrays which you flatMap which involves allocating a bunch of temporary memory.

Reply:
It's been some time but i _think_ i did try sorting the elements and not using `Array(unsafeUninitializedCapacity:)`, but using it was faster IIRC. I remember i speed-tested something in one of these PRs but i don't remember for sure.

Reply:
That's what i was talking about. There is no such thing as `$1.count`. It would be `$1.underestimatedCount`:
```swift
let count = result.reduce(into: 0) { $0 += $1.1.underestimatedCount }
```
`underestimatedCount` is not enough for `Array(unsafeUninitializedCapacity:)`. The documentation mentions the count we provide must be accurate: not more and not less.

Reply:
We could try to pull off a specialized function when the Sequence is an Array, I guess? Doing it for `Collection` could be a little bit tricky if the `count` is not accurate 🤔.

Reply:
Yeah I guess. I assumed the flatMap closure always returned an `Array`.